share/completions/meson.fish

```diff
@@ -1,52 +1,308 @@
-    ",mandir,Manual page directory [share/man]" \
-    ",datadir,Data file directory [share]" \
-    ",bindir,Executable directory [bin]" \
-    ",sharedstatedir,Arch-agnostic data directory [com]" \
-    ",libdir,Library directory [system default]" \
-    ",localstatedir,Localstate data directory [var]" \
-    ",libexecdir,Library executable directory [libexec]" \
-    ",includedir,Header file directory [include]" \
-    ",sysconfdir,Sysconf data directory [etc]"
-
-for arg in $basic_arguments
-    set -l parts (string split , -- $arg)
-    if not string match -q "" -- $parts[1]
-        complete -c meson -s "$parts[1]" -l "$parts[2]" -d "$parts[3]"
+function __fish_meson_needs_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    argparse -s 'v/version' -- $cmd 2>/dev/null
+    or return 0
+    not set -q argv[1]
+end
+
+function __fish_meson_using_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    test (count $cmd) -eq 0
+    and return 1
+    contains -- $cmd[1] $argv
+    and return 0
+end
+
+function __fish_meson_builddir
+    # Consider the value of -C option to detect the build directory
+    set -l cmd (commandline -opc)
+    argparse -i 'C=' -- $cmd
+    if set -q _flag_C
+        echo $_flag_C
     else
-        complete -c meson -l "$parts[2]" -d "$parts[3]"
+        echo .
     end
 end
 
-for arg in $dir_arguments
-    set -l parts (string split , -- $arg)
-    complete -c meson -l "$parts[2]" -d "$parts[3]" -xa '(__fish_complete_directories)'
+function __fish_meson_targets
+    set -l python (__fish_anypython); or return
+    meson introspect --targets (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+targets = set()
+for target in data:
+    targets.add(target["name"])
+for name in targets:
+    print(name)' 2>/dev/null
+end
+
+function __fish_meson_subprojects
+    set -l python (__fish_anypython); or return
+    meson introspect --projectinfo (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+for subproject in data["subprojects"]:
+    print(subproject["name"])' 2>/dev/null
+end
+
+function __fish_meson_tests
+    # --list option shows suites in a short form, e.g. if a test "gvariant"
+    # is present both in "glib:glib" and "glib:slow" suites, it will be shown
+    # in a list as "glib:glib+slow / gvariant". So, just filter out the first
+    # part and list all of the test names.
+    meson test -C (__fish_meson_builddir) --no-rebuild --list | string split -r -f1 ' / '
+end
+
+function __fish_meson_test_suites
+    set -l python (__fish_anypython); or return
+    meson introspect --tests (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+suites = set()
+for test in data:
+    suites.update(test["suite"])
+for name in suites:
+    print(name)' 2>/dev/null
+end
+
+function __fish_meson_help_commands
+    meson help --help | string match -g -r '^  {(.*)}' | string split ,
```

Comment:
Let's be a bit more flexible and support a variable number of spaces, can you change that to 

```
meson help --help | string match -g -r '^ *{(.*)}' | string split ,
```

Reply:
Sure. I have updated the commit.